File: cpp/array/prefixSum.cpp
#include <vector>

using std::vector;

class PrefixSum {
public:
    vector<int> prefix_;

    PrefixSum(vector<int>& nums) {
        int total = 0;
        for (int n : nums) {
            total += n;
            prefix_.push_back(total);
        }
    }

    int rangeSum(int left, int right) {
        int preRight = prefix_[right];
        int preLeft = left > 0 ? prefix_[left - 1] : 0;
        return preRight - preLeft;
    }
};

File: swift/array/prefixSum.swift
class PrefixSum {
  var prefix: [Int] = []

  init(_ nums: [Int]) {
    var total = 0
    for n in nums {
      total += n
      prefix.append(total)
    }
  }

  func rangeSum(_ left: Int, _ right: Int) -> Int {
    let preRight = prefix[right]
    let preLeft = left > 0 ? prefix[left - 1] : 0
    return preRight - preLeft
  }
}